File: smartshop.js
const crypto = require('crypto');

// RSA public key as a string
const publicKey = `-----BEGIN PUBLIC KEY-----
MIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQCRn+iPY/ENsTQpLsyIDPK/HRzv
irt81Wc8Nl9Iv/Vt10hSsefW98j1vo0RaBOYUYpVeSaM13C/r0LqSFkF/gC6t5vr
U3bJ6vLfLg9IDx33h+G5aT78ZHyVdj1VBiJBIQxmd9tV+xphm1dQsptZEzJ2t/0Y
7U7BSRu35ERVxi+HzwIDAQAB
-----END PUBLIC KEY-----`;

// Function to generate current timestamp in the format YYYY-MM-DD HH:mm:ss
function getCurrentTimestamp() {
    const now = new Date();
    const year = now.getFullYear();
    const month = String(now.getMonth() + 1).padStart(2, '0'); // Months are 0-based
    const day = String(now.getDate()).padStart(2, '0');
    const hours = String(now.getHours()).padStart(2, '0');
    const minutes = String(now.getMinutes()).padStart(2, '0');
    const seconds = String(now.getSeconds()).padStart(2, '0');
    return `${year}-${month}-${day} ${hours}:${minutes}:${seconds}`;
}

// Function to generate a random GUID
function generateGUID() {
    return crypto.randomUUID(); // Generates a random UUID (GUID)
}

// Generate random string
const prefix = "ss_android_mobile_1k";
const timestamp = getCurrentTimestamp();
const guid = generateGUID();
const randomString = `${prefix}#${timestamp}#${guid}`;

console.log('Random String:', randomString);

try {
    // Encrypt the random string using the public key with RSA/ECB/PKCS1Padding
    const encryptedBuffer = crypto.publicEncrypt(
        {
            key: publicKey,
            padding: crypto.constants.RSA_PKCS1_PADDING
        },
        Buffer.from(randomString, 'utf-8')
    );

    // Convert the encrypted data to Base64
    const encryptedBase64 = encryptedBuffer.toString('base64');
    console.log('Encrypted (Base64):', encryptedBase64);

    // Combine prefix and encrypted text and Base64 encode the final result
    const finalText = `${prefix}:${encryptedBase64}`;
    const finalBase64 = Buffer.from(finalText).toString('base64');

    console.log('Final Base64:', finalBase64);
} catch (error) {
    console.error('Error encrypting:', error.message);
}

function generateSerial() {
  const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';
  let serial = '';

  // Generate a 12-character alphanumeric string
  for (let i = 0; i < 12; i++) {
    serial += chars.charAt(Math.floor(Math.random() * chars.length));
  }

  return serial;
}

console.log('X-Device-Id:', generateSerial());

File: smartshop.go
package main

import (
	"crypto/rand"
	"crypto/rsa"
	"crypto/x509"
	"encoding/base64"
	"encoding/pem"
	"log"
	"net/http"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
)

const rsaPublicKeyPEM = `-----BEGIN PUBLIC KEY-----
MIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQCRn+iPY/ENsTQpLsyIDPK/HRzv
irt81Wc8Nl9Iv/Vt10hSsefW98j1vo0RaBOYUYpVeSaM13C/r0LqSFkF/gC6t5vr
U3bJ6vLfLg9IDx33h+G5aT78ZHyVdj1VBiJBIQxmd9tV+xphm1dQsptZEzJ2t/0Y
7U7BSRu35ERVxi+HzwIDAQAB
-----END PUBLIC KEY-----`

func main() {
	gin.SetMode(gin.ReleaseMode)

	// Use gin.New() to avoid the default logger middleware that produces [GIN] logs.
	r := gin.New()

	// Add only the recovery middleware to handle panics without logging requests.
	r.Use(gin.Recovery())

	r.GET("/smartshop", func(c *gin.Context) {
		// Generate timestamp in the same format as JavaScript
		now := time.Now().Format("2006-01-02 15:04:05")
		guid := uuid.New().String()
		prefix := "ss_android_mobile_1k"

		// Create the random string in the same format as JavaScript
		randomString := prefix + "#" + now + "#" + guid

		// Load RSA public key
		pubKey, err := loadRSAPublicKeyFromPEM([]byte(rsaPublicKeyPEM))
		if err != nil {
			log.Printf("Error loading public key: %v", err)
			c.String(http.StatusInternalServerError, "Error loading public key")
			return
		}

		// Encrypt the random string using RSA with PKCS1 padding
		encryptedBytes, err := rsa.EncryptPKCS1v15(rand.Reader, pubKey, []byte(randomString))
		if err != nil {
			log.Printf("Error encrypting: %v", err)
			c.String(http.StatusInternalServerError, "Error encrypting text")
			return
		}

		// Convert encrypted data to Base64
		encryptedBase64 := base64.StdEncoding.EncodeToString(encryptedBytes)

		// Combine prefix and encrypted text (matching JavaScript logic)
		finalText := prefix + ":" + encryptedBase64

		// Base64 encode the final result
		finalBase64 := base64.StdEncoding.EncodeToString([]byte(finalText))

		// Return only the final base64 string
		c.String(http.StatusOK, finalBase64)
	})

	// Log server start
	log.Println("Server started on port 3455")

	// Start the server and log fatal errors if any
	if err := r.Run(":4616"); err != nil {
		log.Fatalf("Server exited with error: %v", err)
	}
}

func loadRSAPublicKeyFromPEM(pubPEM []byte) (*rsa.PublicKey, error) {
	block, _ := pem.Decode(pubPEM)
	if block == nil || block.Type != "PUBLIC KEY" {
		return nil, x509.CertificateInvalidError{}
	}

	pubInterface, err := x509.ParsePKIXPublicKey(block.Bytes)
	if err != nil {
		return nil, err
	}

	pubKey, ok := pubInterface.(*rsa.PublicKey)
	if !ok {
		return nil, x509.CertificateInvalidError{}
	}

	return pubKey, nil
}
